test/src/main/java/org/axonframework/test/fixture/AxonTestFixture.java

```diff
@@ -0,0 +1,594 @@
+                                         ? message
+                                         : toGenericEventMessage(e, MetaData.emptyInstance())
+                                 ).toArray(EventMessage<?>[]::new);
+            return events(messages);
+        }
+
+        @Override
+        public AxonTestPhase.Given events(EventMessage<?>... messages) {
+            inUnitOfWorkRunOnInvocation(processingContext -> eventSink.publish(processingContext,
+                                                                               messages));
+            return this;
+        }
+
+        private AsyncUnitOfWork inUnitOfWorkRunOnInvocation(Consumer<ProcessingContext> action) {
+            var unitOfWork = new AsyncUnitOfWork();
+            unitOfWork.runOnInvocation(action);
+            unitsOfWork.add(unitOfWork);
+            return unitOfWork;
+        }
+
+        private AsyncUnitOfWork inUnitOfWorkOnInvocation(Function<ProcessingContext, CompletableFuture<?>> action) {
+            var unitOfWork = new AsyncUnitOfWork();
+            unitOfWork.onInvocation(action);
+            unitsOfWork.add(unitOfWork);
+            return unitOfWork;
+        }
+
+        @Override
+        public AxonTestPhase.Given command(Object payload, MetaData metaData) {
+            var messageType = messageTypeResolver.resolve(payload);
+            var commandMessage = new GenericCommandMessage<>(
+                    messageType,
+                    payload,
+                    metaData
+            );
+            return commands(commandMessage);
+        }
+
+        @Override
+        public AxonTestPhase.Given commands(CommandMessage<?>... messages) {
+            inUnitOfWorkOnInvocation(processingContext -> {
+                CompletableFuture<? extends Message<?>> dispatchOneByOneFuture = CompletableFuture.completedFuture(null);
+                for (var message : messages) {
+                    var dispatchFuture = commandBus.dispatch(message, processingContext);
+                    dispatchOneByOneFuture = dispatchOneByOneFuture.thenCompose(m -> dispatchFuture);
+                }
+                return dispatchOneByOneFuture;
+            });
+            return this;
+        }
+
+        @Override
+        public AxonTestPhase.When when() {
+            for (var unitOfWork : unitsOfWork) {
+                awaitCompletion(unitOfWork.execute());
+            }
+            return new When(configuration, customization, messageTypeResolver, commandBus, eventSink);
+        }
+
+        private void awaitCompletion(CompletableFuture<?> completion) {
+            completion.join();
+        }
+    }
+
+
+    static class When implements AxonTestPhase.When {
+
+        private final NewConfiguration configuration;
+        private final Customization customization;
+        private final MessageTypeResolver messageTypeResolver;
+        private final RecordingCommandBus commandBus;
+        private final RecordingEventSink eventSink;
```

Comment:
Same argument as on the `static class Given`: you could choose to access these fields globally, or you could make this a separate class. I guess I would go the latter approach of the two.